### DISCUSSION TITLE
Federation status command not found via CLI

### DISCUSSION THREAD

**Discussion by sturla78** [rabbitmq/rabbitmq-server#4064] [Other] [answered]
The command 
rabbitmqctl federation_status
return a Command 'federation_status' not found.

I have correctly enable the plug-in, the rabitmq version is 3.8.9 and erlang is 23.1

The alternative web API work correctly but i prefer the command for a script check.

What can i do for understand the problem ?

Thanks

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
For plugin commands to be discovered, a few things have to hold true:

 * Your CLI tools must contact the correct node, use `--node [node@name]` if necessary
 * The plugin must be enabled on said node
 * CLI tools must be able to read that node's `enabled_plugins` file and locate all those plugins locally (the latter should not be an issue for tier 1 plugins such as `rabbitmq_federation`)
 * The command must attach to the right tool (you cannot assume that `rabbitmqctl` exposes every command, even though it does expose `federation_status` specifically)

So my best guess is that either point 1 or 3 are off in your case. It's trivial to see plugins are enabled and inspect the discovered commands using `rabbitmqctl help`, `rabbitmq-diagnostics help` and so on.

RabbitMQ 3.8.9 is from [September 2020](https://www.rabbitmq.com/changelog.html) and is close to 20 patch releases behind even 3.8.

**Comment by sturla78** [comment]
Thanks michaelklishin for arrange for me the discussion, and thanks for help.

I try to write a litle bit of information. I use federation for cross comunication between virtual hosts on rabbitmq, so all configuration, upstream, policy ecc, are on the same machine, i haven't a real remote comunication. 

I run under windows 10, and after installed rabbitmq i have enabled the plugin, then i open the command prompt of rabbitmq using the windows menu, and the console open directly on the correct dir under C:\Program Files\RabbitMQ Server\rabbitmq_server-3.8.9\sbin

then i run "rabbitmq-plugins.bat list"
Listing plugins with pattern ".*" ...
 Configured: E = explicitly enabled; e = implicitly enabled
 | Status: * = running on rabbit@vmErlTest
 |/
[  ] rabbitmq_amqp1_0                  3.8.9
[  ] rabbitmq_auth_backend_cache       3.8.9
[  ] rabbitmq_auth_backend_http        3.8.9
[  ] rabbitmq_auth_backend_ldap        3.8.9
[  ] rabbitmq_auth_backend_oauth2      3.8.9
[  ] rabbitmq_auth_mechanism_ssl       3.8.9
[  ] rabbitmq_consistent_hash_exchange 3.8.9
[  ] rabbitmq_event_exchange           3.8.9
[e*] rabbitmq_federation               3.8.9
[E*] rabbitmq_federation_management    3.8.9
[  ] rabbitmq_jms_topic_exchange       3.8.9
[E*] rabbitmq_management               3.8.9
[e*] rabbitmq_management_agent         3.8.9
[  ] rabbitmq_mqtt                     3.8.9
[  ] rabbitmq_peer_discovery_aws       3.8.9
[  ] rabbitmq_peer_discovery_common    3.8.9
[  ] rabbitmq_peer_discovery_consul    3.8.9
[  ] rabbitmq_peer_discovery_etcd      3.8.9
[  ] rabbitmq_peer_discovery_k8s       3.8.9
[  ] rabbitmq_prometheus               3.8.9
[  ] rabbitmq_random_exchange          3.8.9
[  ] rabbitmq_recent_history_exchange  3.8.9
[  ] rabbitmq_sharding                 3.8.9
[e*] rabbitmq_shovel                   3.8.9
[E*] rabbitmq_shovel_management        3.8.9
[  ] rabbitmq_stomp                    3.8.9
[  ] rabbitmq_top                      3.8.9
[E*] rabbitmq_tracing                  3.8.9
[  ] rabbitmq_trust_store              3.8.9
[e*] rabbitmq_web_dispatch             3.8.9
[  ] rabbitmq_web_mqtt                 3.8.9
[  ] rabbitmq_web_mqtt_examples        3.8.9
[  ] rabbitmq_web_stomp                3.8.9
[  ] rabbitmq_web_stomp_examples       3.8.9

for verify the enabled_plugins file i have run "rabbitmqctl.bat status" (i don't known if this is a correct procedure) but the status show a lot of information including 

Enabled plugin file: c:/Users/Administrator/AppData/Roaming/RabbitMQ/enabled_plugins
Enabled plugins:

 * rabbitmq_federation_management
 * rabbitmq_federation
 * rabbitmq_shovel_management
 * rabbitmq_shovel
 * amqp10_client
 * amqp10_common
 * rabbitmq_tracing
 * rabbitmq_management
 * amqp_client
 * rabbitmq_web_dispatch
 * cowboy
 * cowlib
 * rabbitmq_management_agent

so i have open the c:/Users/Administrator/AppData/Roaming/RabbitMQ/enabled_plugins and inside there is this content:

[rabbitmq_federation_management,rabbitmq_management,rabbitmq_shovel_management,rabbitmq_tracing].

but you are right, there is something that i not understand because both command "federation_status" and "shovel_status" return same error, command not found. 

On the other side if i open the web management console, under admin -> federation status, i can observe that all federation link are running (green state).

**Reply by michaelklishin** [reply]
You haven't actually enabled the federation plugin according to the above output. It was enabled as a dependency of `rabbitmq_federation_management` which does not provide any CLI commands.

**Comment by michaelklishin** [comment, marked as answer]
I can reproduce and this is a known behavior: if you don't enable a plugin explicitly, it won't be included in the [Enabled Plugins File](https://rabbitmq.com/plugins.html#enabled-plugins-file) which CLI tools use to determine what code paths to use to discover commands.

Enabling `rabbitmq_federation` explicitly and restarting the node makes `federation_status` discoverable.

**Comment by sturla78** [comment]
Yess, 
thanks michaelklishin you have point to the right direction, i have re-enable the plugin explicitly and now the command run correctly and i can show the status of federation, thanks.

Anyway for all that are in this situation, searching on internet i have found also an alias (i don't know if this is correct) that work also without install plugin explicitly:

rabbitmqctl eval "rabbit_federation_status:status()."

this work also for shovel.

Again thanks michaelklishin.

**Comment by michaelklishin** [comment]
We discourage the use of `rabbitmqctl eval` unless that's the only option to inspect something.

**Comment by michaelklishin** [comment]
I have pushed a couple of updates to our doc guides, such as the [CLI guide](https://rabbitmq.com/cli.html). They explain that for plugin commands to be discoverable, the plugin must be explicitly enabled. It will take a few minutes for the changes to go live
and get through CDN caches.

**Comment by michaelklishin** [comment]
The new section [is live now](https://rabbitmq.com/cli.html#command-discovery)

**Reply by sturla78** [reply]
Thanks michaelklishin, very appreciated.
